Sources/BraintreePayPal/BTPayPalClient.swift

```diff
@@ -272,27 +276,44 @@ import BraintreeDataCollector
                     self.notifyFailure(with: BTPayPalError.httpPostRequestError(dictionary), completion: completion)
                     return
                 }
-
-                guard let body,
-                      let approvalURL = body["paymentResource"]["redirectUrl"].asURL() ??
-                        body["agreementSetup"]["approvalUrl"].asURL() else {
+                
+                guard let body, let approvalURL = BTPayPalApprovalURLParser(body: body) else {
                     self.notifyFailure(with: BTPayPalError.invalidURL, completion: completion)
                     return
                 }
-
-                let pairingID = self.token(from: approvalURL)
-
-                if !pairingID.isEmpty {
-                    self.payPalContextID = pairingID
-                }
+                
+                self.payPalContextID = approvalURL.pairingID
 
                 let dataCollector = BTDataCollector(apiClient: self.apiClient)
-                self.clientMetadataID = self.payPalRequest?.riskCorrelationID ?? dataCollector.clientMetadataID(pairingID)
-                self.handlePayPalRequest(with: approvalURL, paymentType: request.paymentType, completion: completion)
+                self.clientMetadataID = self.payPalRequest?.riskCorrelationID ?? dataCollector.clientMetadataID(approvalURL.pairingID)
+                
+                switch approvalURL.redirectType {
+                case .payPalApp(let url):
+                    self.launchPayPalApp(with: url, completion: completion)
+                case .webBrowser(let url):
+                    self.handlePayPalRequest(with: url, paymentType: request.paymentType, completion: completion)
+                }
             }
         }
     }
     
+    private func launchPayPalApp(with paypalAppRedirectUrl: URL, completion: @escaping (BTPayPalAccountNonce?, Error?) -> Void) {
+        var urlComponents = URLComponents(url: paypalAppRedirectUrl, resolvingAgainstBaseURL: true)
+        urlComponents?.queryItems = [
+            URLQueryItem(name: "source", value: "braintree_sdk"),
+            URLQueryItem(name: "switch_initiated_time", value: String(UInt64(Date().timeIntervalSince1970 * 1000)))
```

Comment:
Updated! Also using `Int` instead of `UInt64`, which I think we only needed when we supported 32 bit devices way back when

Ran some local tests to confirm the rounding looks good:
* 1711390872156.703 --> 1711390872157
* 1711390872661.359 --> 1711390872661
* 1711390873618.973 --> 1711390873619

Reply:
We'll also want to make that change in `main` for our analytics timestamps

Reply:
DTBTSDK-3703 - ticket to fix on `main`